psh/pshapp/pshapp.c

```diff
@@ -812,20 +813,95 @@ static int psh_readcmd(struct termios *orig, psh_hist_t *cmdhist, char **cmd)
 }
 
 
-static int psh_parsecmd(char *line, int *argc, char ***argv)
+static char *psh_unescape(char *token, char ch)
 {
-	char *cmd, *arg, **rargv;
+	char *src = token;
+	char *dst = token;
 
-	if ((cmd = strtok(line, "\t ")) == NULL)
+	while (*src != '\0') {
+		if (src[0] == '\\' && src[1] == ch) {
+			*(dst++) = ch;
+			src += 2;
+		}
+		else {
+			*(dst++) = *(src++);
+		}
+	}
+	*dst = '\0';
+	return token;
+}
+
+
+static char *psh_tokenize(char **pstr)
+{
+	char *ptr = *pstr;
+	char *ret = *pstr;
+	char quote = 0;
+
+	if (*ptr == '\0') {
+		return NULL;
+	}
+
+	while (*ptr != '\0' && isspace(*ptr)) {
+		ptr++;
+	}
+
+	ret = ptr;
+
+	if (*ptr == '"' || *ptr == '\'') {
+		quote = *(ptr++);
+		ret = ptr;
+		for (;;) {
+			while (*ptr != '\0' && *ptr != quote) {
+				ptr++;
+			}
+			if (*ptr == '\0' || *(ptr - 1) != '\\') {
+				break;
+			}
```

Comment:
This can be added to skip the `psh_unescape()` function call if there is no backaslash in quotation marks.
On top declare:
```c
int need_unscape = 0;
```
```suggestion
			if (*ptr == '\0') {
				break;
			}
			if (*(ptr - 1) == '\\') {
				need_unscape = 1;
			}
			else {
				break;
			}
```
and return
```c
return (need_unscape == 0) ? ret : psh_unescape(ret, quote);
```